data/service/api/v1/devicetokens.go

```diff
@@ -0,0 +1,45 @@
+package v1
+
+import (
+	data "github.com/tidepool-org/platform/data/service"
+	"github.com/tidepool-org/platform/devicetokens"
+	"github.com/tidepool-org/platform/request"
+	platform "github.com/tidepool-org/platform/service"
+	"github.com/tidepool-org/platform/service/api"
+)
+
+func DeviceTokensRoutes() []data.Route {
+	return []data.Route{
+		data.MakeRoute("POST", "/v1/device-tokens/:userID", UpsertDeviceToken, api.RequireAuth),
+	}
+}
+
+func UpsertDeviceToken(dCtx data.Context) {
+	r := dCtx.Request()
+	ctx := r.Context()
+	authDetails := request.GetAuthDetails(ctx)
+	repo := dCtx.DeviceTokensRepository()
+
+	if err := checkAuthentication(authDetails); err != nil {
+		dCtx.RespondWithError(platform.ErrorUnauthorized())
+		return
+	}
+
+	if err := checkUserIDConsistency(authDetails, r.PathParam("userID")); err != nil {
+		dCtx.RespondWithError(platform.ErrorUnauthorized())
+		return
+	}
+
+	deviceToken := devicetokens.DeviceToken{}
```

Comment:
Validation of `deviceToken`? We don't want bogus data in the database.

Reply:
Is there actually a way to validate the device token? Other than sending a test message that is?

I think it's supposed to be entirely opaque blob. At most I guess we could cap it at N characters / bytes?

Reply:
@tjotala At this point in the workflow, we cannot ensure that the device token is truly valid (and will actually "connect" to a physical device), but we can at least ensure it passes some level of validation. In particular, I'm thinking of the code already in the `Validate` function on the `DeviceToken` struct.

Reply:
I see no means of validating. If we really want to, I guess we could set some ridiculously high length setting, but even that goes against Apple's guidelines:

> deviceToken
>
>    A globally unique token that identifies this device to APNs. Send this token to the server that you use to generate remote notifications. Your server must pass this token unmodified back to APNs when sending those remote notifications.
>
>    APNs device tokens are of variable length. Do not hard-code their size.
>
> https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622958-application#parameters

If you guys can decide what you'd like, I can implement it, but it doesn't seem like anything that should hold up the PR?

Reply:
I think all we can do is to have some reasonably high size limit that a) a legit token will never hit and b) is small enough to discourage script kiddies.

According to https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification and https://firebase.google.com/docs/cloud-messaging/concept-options, the max notification payloads are on the order of 4-5KB, though it's not clear that includes the token. Meanwhile, IIRC most web servers cap HTTP headers at around 8KB to 16KB. And then finally MongoDB caps the max document size at 16MB.

So if we cap the device token size at, say, 8KB, that should be more than sufficient to cover any plausible token, yet far too small for it to be useful for anything sneaky like trying to overflow HTTP headers or MongoDB documents.

Reply:
That sounds reasonable to me. I'll implement it, and we can change it later if we want.

Reply:
Actually, in writing up the docs, I think we can go WAAAAY smaller and still be totally ok... But I'm leaving it at 8 KB for now. It's easy to update later.

Reply:
@ewollesen Much of the "validation" in `platform` does not actually ensure that the data is truly valid, but rather that the data is NOT knowingly invalid. (I suppose a more descriptive name for the `Validate` functions would have been `IsNotInvalid`, but that's a mouthful and the double negative confusing.)

For example, a blood sugar value of 1 mg/dL from a CGM is really not possible (no known CGM returns that low a value nor is it even physiologically possible), but as we don't know of a hard minimum value we can test against, we instead choose a reasonable value of 0 where anything less is definitely invalid (a negative number).

In any case, what I was suggest was simply invoking the `DeviceToken.Validate` function which already does validation.